### PR TITLE
ARROW-6035: [Java] Avro adapter support convert nullable value

### PR DESCRIPTION
**Pull request by tianchen92**
Related to [ARROW-6035](https://issues.apache.org/jira/browse/ARROW-6035).

A  specific Avro unions type(has two types and one is null type) could convert to a nullable ArrowVector.
For instance, ["null", "string"] could represented by a VarcharVector which could has null value.

### CODE REVIEW

File: java/adapter/avro/src/main/java/org/apache/arrow/AvroToArrowUtils.java [hunk 1]
```diff
@@ -75,29 +76,61 @@
...
+    long nullCount = schema.getTypes().stream().filter(s -> s.getType() == Type.NULL).count();
+    // if has two field and one is null type, convert to nullable primitive type
+    if (size == 2 && nullCount == 1) {
+      Schema subSchema = schema.getTypes().stream().filter(s -> s.getType() != Type.NULL).findFirst().get();
+      String nullIndex = schema.getTypes().indexOf(subSchema) == 0 ? "1" : "0";
+      subSchema.addProp(NULL_INDEX, nullIndex);
+      Preconditions.checkNotNull(subSchema);
+      return getArrowField(subSchema, name,true);
+    } else {
+      //TODO convert avro unions type to arrow UnionVector
```
Comment (emkornfield) [inline java/adapter/avro/src/main/java/org/apache/arrow/AvroToArrowUtils.java]: throw a NotSupportedException instead.

File: java/adapter/avro/src/main/java/org/apache/arrow/AvroToArrowUtils.java [hunk 2]
```diff
@@ -75,29 +76,61 @@
...
+    final FieldType fieldType =  new FieldType(/*nullable=*/nullable, arrowType, /*dictionary=*/null,
+        /*metadata=*/getMetaData(schema));
+    return new Field(name, fieldType, null);
+  }
+
+  private static Field getUnionField(Schema schema, String name) {
+    int size = schema.getTypes().size();
+    long nullCount = schema.getTypes().stream().filter(s -> s.getType() == Type.NULL).count();
+    // if has two field and one is null type, convert to nullable primitive type
+    if (size == 2 && nullCount == 1) {
```
Comment (emkornfield) [inline java/adapter/avro/src/main/java/org/apache/arrow/AvroToArrowUtils.java]: can types repeat in Avro?  Should this be size == nullCount + 1

File: java/adapter/avro/src/main/java/org/apache/arrow/consumers/AvroIntConsumer.java [hunk 3]
```diff
@@ -21,24 +21,37 @@
...
+    this.nullable = vector.getField().isNullable();
+    if (nullable) {
+      getNullFieldIndex(vector.getField());
+    }
   }
 
   @Override
   public void consume(Decoder decoder) throws IOException {
-    writer.writeInt(decoder.readInt());
+    if (!nullable) {
```
Comment (emkornfield) [inline java/adapter/avro/src/main/java/org/apache/arrow/consumers/AvroIntConsumer.java]: did you consider using decomposition?  Something like:
```
class NullableTypeConsumer {
   Consumer delegate;
   int nullIndex = 0;
   @Override 
   public void consume(Decoder decoder) {
       if (nullIndex != decoder.readInt()) {
           delegate.consume(decoder);
       } else {
          writer.setPosition(writer.getPosition() + 1)
       }
      
   }
}
```

File: java/adapter/avro/src/main/java/org/apache/arrow/consumers/AvroIntConsumer.java [hunk 4]
```diff
@@ -21,24 +21,37 @@
...
 
 /**
  * Consumer which consume int type values from avro decoder.
  * Write the data to {@link IntVector}.
  */
-public class AvroIntConsumer implements Consumer {
+public class AvroIntConsumer extends Consumer {
 
-  private final IntWriter writer;
+  private final IntWriterImpl writer;
```
Comment (emkornfield) [inline java/adapter/avro/src/main/java/org/apache/arrow/consumers/AvroIntConsumer.java]: why this change?

File: java/adapter/avro/src/main/java/org/apache/arrow/AvroToArrowUtils.java (hunk 1 above)
Comment (tianchen92) [reply java/adapter/avro/src/main/java/org/apache/arrow/AvroToArrowUtils.java]: Fixed.

File: java/adapter/avro/src/main/java/org/apache/arrow/AvroToArrowUtils.java (hunk 2 above)
Comment (tianchen92) [reply java/adapter/avro/src/main/java/org/apache/arrow/AvroToArrowUtils.java]: Repeated types are not allowed. Right, size == nullCount + 1 could check nullable and size == 2 could check nullable primitive type, fixed.

File: java/adapter/avro/src/main/java/org/apache/arrow/consumers/AvroIntConsumer.java (hunk 3 above)
Comment (tianchen92) [reply java/adapter/avro/src/main/java/org/apache/arrow/consumers/AvroIntConsumer.java]: Good suggestion, this will be really helpful in Union complex types. 👍

File: java/adapter/avro/src/main/java/org/apache/arrow/consumers/AvroIntConsumer.java (hunk 4 above)
Comment (tianchen92) [reply java/adapter/avro/src/main/java/org/apache/arrow/consumers/AvroIntConsumer.java]: Not meant to do this, reverted, thanks for reminder.

**Comment by emkornfield** [comment]
Sorry, will take a look tomorrow at the updates.

File: java/adapter/avro/src/main/java/org/apache/arrow/consumers/NullablePrimitiveTypeConsumer.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+import java.io.IOException;
+
+import org.apache.avro.io.Decoder;
+
+/**
+ * Consumer holds a primitive consumer, could consume nullable values from avro decoder.
+ * Write data via writer of delegate consumer.
+ */
+public class NullablePrimitiveTypeConsumer implements Consumer {
```
Comment (emkornfield) [inline java/adapter/avro/src/main/java/org/apache/arrow/consumers/NullablePrimitiveTypeConsumer.java]: Would this not work for complex types?

File: java/adapter/avro/src/main/java/org/apache/arrow/consumers/Consumer.java [hunk 6]
```diff
@@ -22,9 +22,11 @@
...
 
 /**
- * An abstraction that is used to consume values from avro decoder.
+ * Interface that is used to consume values from avro decoder.
  */
 public interface Consumer {
 
   void consume(Decoder decoder) throws IOException;
+
+  void movePosition();
```
Comment (emkornfield) [inline java/adapter/avro/src/main/java/org/apache/arrow/consumers/Consumer.java]: Maybe something more literal like "addNull" this makes it clear what is being done.  

In either case please add documents for the intentended contract of each method.

**Review by emkornfield — CHANGES_REQUESTED** [review]
Looks much nicer with delegation.  Thank you.  Still a few more points that need to be addressed.

File: java/adapter/avro/src/main/java/org/apache/arrow/consumers/NullablePrimitiveTypeConsumer.java (hunk 5 above)
Comment (tianchen92) [reply java/adapter/avro/src/main/java/org/apache/arrow/consumers/NullablePrimitiveTypeConsumer.java]: I think it won't work for complex types, for unions type like ["string", "int", "null"] I think should add a consumer which holds something like Map<index, delegate>.
This PR only aims for nullable primitive type represented by union type.

File: java/adapter/avro/src/main/java/org/apache/arrow/consumers/Consumer.java (hunk 6 above)
Comment (tianchen92) [reply java/adapter/avro/src/main/java/org/apache/arrow/consumers/Consumer.java]: Sure, fixed.

File: java/adapter/avro/src/main/java/org/apache/arrow/AvroToArrowUtils.java [hunk 7]
```diff
@@ -75,29 +78,70 @@
...
       case BYTES:
-        return new ArrowType.Binary();
+        arrowType = new ArrowType.Binary();
+        break;
       default:
         // no-op, shouldn't get here
         throw new RuntimeException("Can't convert avro type %s to arrow type." + type.getName());
     }
+
+    final FieldType fieldType =  new FieldType(/*nullable=*/nullable, arrowType, /*dictionary=*/null,
```
[the remark this answers is not in the document]
Comment (tianchen92) [reply java/adapter/avro/src/main/java/org/apache/arrow/AvroToArrowUtils.java]: Thanks for reminder, fixed.

File: java/adapter/avro/src/main/java/org/apache/arrow/consumers/NullablePrimitiveTypeConsumer.java (hunk 5 above)
Comment (emkornfield) [reply java/adapter/avro/src/main/java/org/apache/arrow/consumers/NullablePrimitiveTypeConsumer.java]: Agreed, by complex I mean something like [Struct, Null] or [List, Null].  I think a better name would be NullableTypeConsumer?

Comment (tianchen92) [reply java/adapter/avro/src/main/java/org/apache/arrow/consumers/NullablePrimitiveTypeConsumer.java]: Ah, right, fixed.

**Comment by tianchen92** [comment]
@emkornfield Build passed, I think we could get this merged :)

**Comment by emkornfield** [comment]
yep. +1, thank you.
